src/middlewared/middlewared/plugins/keychain.py

```diff
@@ -0,0 +1,376 @@
+    ))
+    def remote_ssh_semiautomatic_setup(self, data):
+        replication_key = self.middleware.run_coroutine(
+            get_ssh_key_pair_with_private_key(self.middleware, data["private_key"]))
+
+        try:
+            client = Client(os.path.join(re.sub("^http", "ws", data["url"]), "websocket"))
+        except Exception as e:
+            raise CallError(f"Unable to connect to remote system: {e}")
+
+        with client as c:
+            if not c.call("auth.token", data["token"]):
+                raise CallError("Invalid token")
+
+            try:
+                response = c.call("replication.pair", {
+                    "hostname": "any-host",
+                    "public-key": replication_key["attributes"]["public_key"],
+                    "user": data["username"],
+                })
+            except Exception as e:
+                raise CallError(f"Semi-automatic SSH connection setup failed: {e!r}")
+
+        return self.middleware.call_sync("keychaincredential.create", {
+            "name": data["name"],
+            "type": "SSH_CREDENTIALS",
+            "attributes": {
+                "host": urllib.parse.urlparse(data["url"]).hostname,
+                "port": response["ssh_port"],
+                "username": data["username"],
+                "private_key": replication_key["id"],
+                "remote_host_key": process_ssh_keyscan_output(response["ssh_hostkey"]),
+                "cipher": data["cipher"],
+                "connect_timeout": data["connect_timeout"],
+            }
+        })
+
+    @private
+    @accepts(Dict(
+        "keychain_ssh_pair",
+        Str("remote_hostname", required=True),
+        Str("username", default="root"),
+        Str("public_key", required=True),
+    ))
+    async def ssh_pair(self, data):
+        """
+        Receives public key, storing it to accept SSH connection and return
+        pertinent SSH data of this machine.
+        """
+        service = await self.middleware.call(
+            "datastore.query", "services.services", [("srv_service", "=", "ssh")], {"get": True})
+        ssh = await self.middleware.call("datastore.query", "services.ssh", None, {"get": True})
+        try:
+            user = await self.middleware.call(
+                "datastore.query", "account.bsdusers", [("bsdusr_username", "=", data["username"])],
+                {"get": True})
+        except IndexError:
+            raise CallError(f"User {data['username']} does not exist")
+
+        if user["bsdusr_home"].startswith("/nonexistent") or not os.path.exists(user["bsdusr_home"]):
+            raise CallError(f"Home directory {user['bsdusr_home']} does not exist", errno.ENOENT)
+
+        # Make sure SSH is enabled
+        if not service["srv_enable"]:
+            await self.middleware.call("datastore.update", "services.services", service["id"], {"srv_enable": True})
+            await self.middleware.call("notifier.start", "ssh")
```

Comment:
Can we use `service.start` now? notifier.start was a backward-compatible thing at the time, now it will just circle back and call middleware using the client.